File: test/shared/logic/play_validator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:koutbh/shared/models/card.dart';
import 'package:koutbh/shared/logic/play_validator.dart';

void main() {
  group('PlayValidator.validatePlay', () {
    test('allows playing a card of the led suit', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(card: GameCard(suit: Suit.hearts, rank: Rank.ace), hand: hand, ledSuit: Suit.hearts, isLeadPlay: false);
      expect(result.isValid, true);
    });

    test('rejects off-suit when player has led suit', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(card: GameCard(suit: Suit.spades, rank: Rank.king), hand: hand, ledSuit: Suit.hearts, isLeadPlay: false);
      expect(result.isValid, false);
      expect(result.error, 'must-follow-suit');
    });

    test('allows off-suit when void in led suit', () {
      final hand = [GameCard(suit: Suit.spades, rank: Rank.king), GameCard(suit: Suit.clubs, rank: Rank.queen)];
      final result = PlayValidator.validatePlay(card: GameCard(suit: Suit.spades, rank: Rank.king), hand: hand, ledSuit: Suit.hearts, isLeadPlay: false);
      expect(result.isValid, true);
    });

    test('allows joker when void in led suit', () {
      final hand = [GameCard.joker(), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(card: GameCard.joker(), hand: hand, ledSuit: Suit.hearts, isLeadPlay: false);
      expect(result.isValid, true);
    });

    test('allows joker even when player has led suit', () {
      final hand = [GameCard.joker(), GameCard(suit: Suit.hearts, rank: Rank.seven)];
      final result = PlayValidator.validatePlay(card: GameCard.joker(), hand: hand, ledSuit: Suit.hearts, isLeadPlay: false);
      expect(result.isValid, true);
    });

    test('rejects leading with joker', () {
      final hand = [GameCard.joker(), GameCard(suit: Suit.hearts, rank: Rank.ace)];
      final result = PlayValidator.validatePlay(card: GameCard.joker(), hand: hand, ledSuit: null, isLeadPlay: true);
      expect(result.isValid, false);
      expect(result.error, 'joker-cannot-lead');
    });

    test('allows leading with any non-joker card', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(card: GameCard(suit: Suit.hearts, rank: Rank.ace), hand: hand, ledSuit: null, isLeadPlay: true);
      expect(result.isValid, true);
    });

    test('kout lead: rejects non-trump when player has trump', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(
        card: GameCard(suit: Suit.spades, rank: Rank.king),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: true,
        isFirstTrick: true,
      );
      expect(result.isValid, false);
      expect(result.error, 'must-lead-trump');
    });

    test('kout lead: allows trump card', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(
        card: GameCard(suit: Suit.hearts, rank: Rank.ace),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: true,
        isFirstTrick: true,
      );
      expect(result.isValid, true);
    });

    test('kout lead: allows any card when void in trump', () {
      final hand = [GameCard(suit: Suit.spades, rank: Rank.king), GameCard(suit: Suit.clubs, rank: Rank.queen)];
      final result = PlayValidator.validatePlay(
        card: GameCard(suit: Suit.spades, rank: Rank.king),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: true,
        isFirstTrick: true,
      );
      expect(result.isValid, true);
    });

    test('kout lead: rejects joker even when holding trump', () {
      final hand = [GameCard.joker(), GameCard(suit: Suit.hearts, rank: Rank.ace)];
      final result = PlayValidator.validatePlay(
        card: GameCard.joker(),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: true,
        isFirstTrick: true,
      );
      expect(result.isValid, false);
      expect(result.error, 'joker-cannot-lead');
    });

    test('kout lead on non-first trick: allows any card', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(
        card: GameCard(suit: Suit.spades, rank: Rank.king),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: true,
        isFirstTrick: false,
      );
      expect(result.isValid, true);
    });

    test('non-kout lead: no trump restriction', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace), GameCard(suit: Suit.spades, rank: Rank.king)];
      final result = PlayValidator.validatePlay(
        card: GameCard(suit: Suit.spades, rank: Rank.king),
        hand: hand,
        ledSuit: null,
        isLeadPlay: true,
        trumpSuit: Suit.hearts,
        isKout: false,
      );
      expect(result.isValid, true);
    });

    test('rejects card not in hand', () {
      final hand = [GameCard(suit: Suit.hearts, rank: Rank.ace)];
      final result = PlayValidator.validatePlay(card: GameCard(suit: Suit.spades, rank: Rank.king), hand: hand, ledSuit: null, isLeadPlay: true);
      expect(result.isValid, false);
      expect(result.error, 'card-not-in-hand');
    });
  });

  group('PlayValidator.playableForCurrentTrick — Joker leading', () {
    test('joker excluded from playable cards when leading', () {
      final hand = [
        GameCard.joker(),
        GameCard(suit: Suit.hearts, rank: Rank.ace),
        GameCard(suit: Suit.spades, rank: Rank.king),
      ];
      final playable = PlayValidator.playableForCurrentTrick(
        hand: hand,
        trickHasNoPlaysYet: true,
        ledSuit: null,
        bidIsKout: false,
        noTricksCompletedYet: false,
      );
      expect(playable, isNot(contains(GameCard.joker())));
      expect(playable.length, 2);
    });

    test('joker allowed when following', () {
      final hand = [
        GameCard.joker(),
        GameCard(suit: Suit.hearts, rank: Rank.ace),
      ];
      final playable = PlayValidator.playableForCurrentTrick(
        hand: hand,
        trickHasNoPlaysYet: false,
        ledSuit: Suit.hearts,
        bidIsKout: false,
        noTricksCompletedYet: false,
      );
      expect(playable, contains(GameCard.joker()));
    });

    test('kout first trick: only trump playable, joker not an option for leading', () {
      final hand = [
        GameCard.joker(),
        GameCard(suit: Suit.hearts, rank: Rank.ace),
        GameCard(suit: Suit.spades, rank: Rank.king),
      ];
      final playable = PlayValidator.playableForCurrentTrick(
        hand: hand,
        trickHasNoPlaysYet: true,
        ledSuit: null,
        trumpSuit: Suit.hearts,
        bidIsKout: true,
        noTricksCompletedYet: true,
      );
      expect(playable, {GameCard(suit: Suit.hearts, rank: Rank.ace)});
    });

    test('only joker in hand when must lead returns empty playable set (poison joker)', () {
      final hand = [GameCard.joker()];
      final playable = PlayValidator.playableForCurrentTrick(
        hand: hand,
        trickHasNoPlaysYet: true,
        ledSuit: null,
        bidIsKout: false,
        noTricksCompletedYet: false,
      );
      expect(playable, isEmpty);
      expect(PlayValidator.detectPoisonJoker(hand), true);
    });
  });

  group('PlayValidator.detectPoisonJoker', () {
    test('detects poison joker when only card is joker', () {
      expect(PlayValidator.detectPoisonJoker([GameCard.joker()]), true);
    });

    test('no poison joker with multiple cards', () {
      expect(PlayValidator.detectPoisonJoker([GameCard.joker(), GameCard(suit: Suit.hearts, rank: Rank.ace)]), false);
    });

    test('no poison joker when single card is not joker', () {
      expect(PlayValidator.detectPoisonJoker([GameCard(suit: Suit.hearts, rank: Rank.ace)]), false);
    });

    test('no poison joker with empty hand', () {
      expect(PlayValidator.detectPoisonJoker([]), false);
    });
  });
}

File: workers/test/game/play-validator.test.ts
import { describe, it, expect } from 'vitest';
import { validatePlay, detectPoisonJoker } from '../../src/game/play-validator';

describe('validatePlay', () => {
  it('rejects card not in hand', () => {
    expect(validatePlay('SA', ['SK', 'HK'], null, true).valid).toBe(false);
  });

  it('accepts card in hand when leading', () => {
    expect(validatePlay('SA', ['SA', 'HK'], null, true).valid).toBe(true);
  });

  it('rejects Joker lead', () => {
    const result = validatePlay('JO', ['JO', 'SA'], null, true);
    expect(result.valid).toBe(false);
    expect(result.error).toBe('joker-cannot-lead');
  });

  it('allows Joker lead when only card in hand', () => {
    // Joker as sole card still can't lead — this triggers poison joker path in game-room
    const result = validatePlay('JO', ['JO'], null, true);
    expect(result.valid).toBe(false);
    expect(result.error).toBe('joker-cannot-lead');
  });

  it('must follow led suit when able', () => {
    const result = validatePlay('HK', ['HK', 'SA'], 'spades', false);
    expect(result.valid).toBe(false);
    expect(result.error).toBe('must-follow-suit');
  });

  it('allows any card when void in led suit', () => {
    expect(validatePlay('HK', ['HK', 'CA'], 'spades', false).valid).toBe(true);
  });

  it('allows Joker when following suit (exempt from must-follow)', () => {
    expect(validatePlay('JO', ['JO', 'SA'], 'spades', false).valid).toBe(true);
  });

  it('Kout first trick: must lead trump if have it', () => {
    const result = validatePlay('HA', ['HA', 'SA'], 'spades', true, 'spades', true, true);
    expect(result.valid).toBe(false);
    expect(result.error).toBe('must-lead-trump');
  });

  it('Kout first trick: allows trump lead', () => {
    expect(validatePlay('SA', ['SA', 'HA'], 'spades', true, 'spades', true, true).valid).toBe(true);
  });

  it('Kout first trick: allows non-trump if no trump in hand', () => {
    expect(validatePlay('HA', ['HA', 'CA'], 'spades', true, 'spades', true, true).valid).toBe(true);
  });

  it('Kout non-first trick: no trump lead restriction', () => {
    expect(validatePlay('HA', ['HA', 'SA'], 'spades', true, 'spades', true, false).valid).toBe(true);
  });

  it('non-Kout: no trump lead restriction', () => {
    expect(validatePlay('HA', ['HA', 'SA'], 'spades', true, 'spades', false, true).valid).toBe(true);
  });

  it('no led suit (first play or Joker led): any card valid', () => {
    expect(validatePlay('HA', ['HA', 'SA'], null, false).valid).toBe(true);
  });
});

describe('detectPoisonJoker', () => {
  it('detects single Joker in hand', () => {
    expect(detectPoisonJoker(['JO'])).toBe(true);
  });

  it('returns false for Joker with other cards', () => {
    expect(detectPoisonJoker(['JO', 'SA'])).toBe(false);
  });

  it('returns false for single non-Joker', () => {
    expect(detectPoisonJoker(['SA'])).toBe(false);
  });

  it('returns false for empty hand', () => {
    expect(detectPoisonJoker([])).toBe(false);
  });
});
